snapcraft/internal/build_providers/_snap.py

```diff
@@ -0,0 +1,296 @@
+        self._snap_dir_unmounter = snap_dir_unmounter
+        self._file_pusher = file_pusher
+
+        self._registry_data = dict()  # type: Union[None, Dict[str, List[Any]]]
+
+    def _load_registry(self):
+        if self._registry_filepath is None or self._registry_data:
+            return
+        if not os.path.exists(self._registry_filepath):
+            return
+
+        with open(self._registry_filepath) as registry_file:
+            self._registry_data = yaml.load(registry_file)
+
+    def _dump_registry(self):
+        if self._registry_filepath is None or self._registry_data is None:
+            return
+
+        dirpath = os.path.dirname(self._registry_filepath)
+        if dirpath:
+            os.makedirs(dirpath, exist_ok=True)
+
+        with open(self._registry_filepath, "w") as registry_file:
+            yaml.dump(self._registry_data, stream=registry_file)
+
+    @contextlib.contextmanager
+    def _mounted_dir(self):
+        if any((s.get_op() == _SnapOp.INJECT for s in self._snaps)):
+            self._snap_dir_mounter()
+            try:
+                yield
+            finally:
+                self._snap_dir_unmounter()
+        else:
+            yield
+
+    def _disable_and_wait_for_refreshes(self):
+        # Disable autorefresh for 15 minutes,
+        # https://github.com/snapcore/snapd/pull/5436/files
+        now_plus_15 = datetime.datetime.now() + datetime.timedelta(minutes=15)
+        self._runner(
+            [
+                "sudo",
+                "snap",
+                "set",
+                "core",
+                "refresh.hold={}Z".format(now_plus_15.isoformat()),
+            ]
+        )
+        # Auto refresh may have kicked in while setting the hold.
+        logger.debug("Waiting for pending snap auto refreshes.")
+        with contextlib.suppress(errors.ProviderExecError):
+            self._runner(
+                ["sudo", "snap", "watch", "--last=auto-refresh"], hide_output=True
+            )
+
+    def _inject_assertions(self) -> None:
+        assertions = []  # type: List[List[str]]
+        for snap in self._snaps:
+            assertions.extend(snap.get_assertions())
+        if assertions:
+            assertions = [_STORE_ASSERTION] + assertions
+
+        # Return early if not assertions need acking
+        if not assertions:
+            return
+
+        with tempfile.NamedTemporaryFile() as assertion_file:
+            for assertion in assertions:
+                assertion_file.write(repo.snaps.get_assertion(assertion))
+                assertion_file.write(b"\n")
+            assertion_file.flush()
+
+            self._file_pusher(
+                source=assertion_file.name, destination=assertion_file.name
+            )
+            self._runner(["sudo", "snap", "ack", assertion_file.name])
+
+    def _get_latest_revision(self, snap_name):
+        self._load_registry()
+        try:
+            return self._registry_data[snap_name][-1]["revision"]
```

Comment:
Rather than having an explicit `_load_registry()` function as well as a private `_registry_data` attribute, this usage seems better suited to simple lazy evaluation of the property (or a `get_registry()` function, if you prefer) to encapsulate the loading and returning of the registry.

Reply:
I sort of started to dislike properties when I do not need them (like replacing an already in use attribute). They are hard to mock and this clearly does something under the hood. Also,
> Explicit is better than implicit" is in direct conflict with using a property. (It looks like a simple assignment, but it calls a function).

After that philosophical point, if you still want me to change, I will :-)

Reply:
Oh, but I am down with get_registry, I had that originally and was passing around too many params for my taste, but I prefer that indeed

Reply:
I like the symmetry on this one and not sure I understand the change now that I am re-reading.